8272773: Configurable card table card size

vish-chan opened the pull request:
Hi,

Please review the changes to make CardTable entry size configurable. The changes primarily consists of:

1. Addition of a cmdline flag **GCCardSizeInBytes** to make the card size startup time configurable.
2. Setting the card size based on the flag in G1, Parallel and Serial GC memory initialization paths.
3. Setting BlockOffsetTable size and ObjectStartArray size based on the card size.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8272773](https://bugs.openjdk.java.net/browse/JDK-8272773): Configurable card table card size


### Reviewers
 * [Thomas Schatzl](https://openjdk.java.net/census#tschatzl) (@tschatzl - **Reviewer**)
 * [Albert Mingkun Yang](https://openjdk.java.net/census#ayang) (@albertnetymk - **Reviewer**) ⚠️ Review applies to 6d7b0844072105f3e365c1e4d63f1f55d5676a1a


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5838/head:pull/5838` \
`$ git checkout pull/5838`

Update a local copy of the PR: \
`$ git checkout pull/5838` \
`$ git pull https://git.openjdk.java.net/jdk pull/5838/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5838`

View PR using the GUI difftool: \
`$ git pr show -t 5838`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5838.diff">https://git.openjdk.java.net/jdk/pull/5838.diff</a>

</details>

tschatzl commented:
@vish-chan : the change for PR#5909 has finally been integrated. Sorry for the delay, things like a related bug delayed its integration more than expected.

vish-chan commented:
@tschatzl I tested PR#5909 and this change: The combination of 1024b card/512M region is giving 2-3% improvement over 1024b card/32M region in SPECjbb2015 on my setup. I haven't analyzed this is detail though. 

Shall I update the pull request with new patch by:
1. Removing min_card_size code as it is not relevant now.
2. Update minimum card size supported from 512 to 128/256? Need your input on the min card size value and default card size value.

kimbarrett commented on src/hotspot/share/gc/parallel/objectStartArray.hpp:
```diff
@@ -52,11 +52,12 @@ class ObjectStartArray : public CHeapObj<mtGC> {
-    block_size                   = 1 << block_shift,
-    block_size_in_words          = block_size / sizeof(HeapWord)
-  };
+  static uint block_shift;
+  static uint block_size;
+  static uint block_size_in_words;
```
Not a review (yet), just a response to these questions.

Public data members are very rare in HotSpot code.  There are a few POD/standard layout classes (examples include G1BufferNodeList and OopStorage::BasicParState::IterationData), but otherwise accessor functions are preferred.  The style-guide says
"Use public accessor functions for member variables accessed outside the class."

The style-guide says this:
"Class data member names have a leading underscore, and use lowercase with words separated by a single underscore (_foo_bar)."
It makes no distinction for public or not, or for static or not.  That is intentional, based on some discussion about that question when I was updating the style-guide a while ago.

Of course, there are probably counter-examples to both of these lurking in the code base.

tschatzl commented:
> @tschatzl I tested PR#5909 and this change: The combination of 1024b card/512M region is giving 2-3% improvement over 1024b card/32M region in SPECjbb2015 on my setup. I haven't analyzed this is detail though.

Good that the changed did not optimize away the improvements.

> 
> Shall I update the pull request with new patch by:
> 
>     1. Removing min_card_size code as it is not relevant now.

Please do. It is unnecessary now as there are no limitations of the combination of card size/region size for G1 any more.

> 
>     2. Update minimum card size supported from 512 to 128/256? Need your input on the min card size value and default card size value.

I'd say make the minimum default. I looked a bit through literature, and the suggestion is that the minimum size used seems 128 bytes. That might be useful for smallish heaps.

Thanks,
  Thomas

tschatzl commented on src/hotspot/share/gc/parallel/objectStartArray.hpp:
```diff
@@ -52,11 +52,16 @@ class ObjectStartArray : public CHeapObj<mtGC> {
+  static uint block_size;
+  static uint block_size_in_words;
+
+  // This maximum is derived from that we need an extra bit for possible
+  // offsets in the byte for backskip values (this is a hard limit)
```
Maybe it is interesting to make this size depend on current `ObjectAlignmentInBytes` after all. I've seen some recent testing where it proved useful. However that may be done as a separate change.
It also seems that reducing the size of the BOT/card table (increasing this maximum) has diminishing returns e.g. wrt to Card Table Clear times.

Maybe it is interesting for you to try out even larger card sizes though (e.g. an `ObjectAlignmentInBytes` value of `16` should allow a card size of 2048 and so on).

vish-chan replied:
Sure. I'll explore this and share the results.

vish-chan commented:
@tschatzl I've updated the PR with a new patch: Removed the dependency of G1 heap region and card size. Set card size in [128. 1024] with 512 as default.

tschatzl commented on src/hotspot/share/gc/shared/cardTable.cpp:
```diff
@@ -26,13 +26,40 @@
 #include "memory/virtualspace.hpp"
 #include "runtime/java.hpp"
 #include "runtime/os.hpp"
+#include "runtime/globals_extension.hpp"
```
I think we typically include the globals include that contains the flag - this should probably be `#include "gc/shared/gc_globals.hpp` (and sorted correctly in this list).

tschatzl commented on src/hotspot/share/gc/shared/cardTable.cpp:
```diff
@@ -26,13 +26,40 @@
 #include "gc/shared/collectedHeap.hpp"
 #include "gc/shared/space.inline.hpp"
+#include "gc/shared/gcLogPrecious.hpp"
+#include "gc/parallel/objectStartArray.hpp"
```
This include is in the wrong order and should probably be guarded by `#if INCLUDE_PARALLELGC`.

tschatzl commented:
Hi @vish-chan - I changed the name of the CR to "Configurable card table card size" because we are done with the investigation :) No, honestly, "investigate xyz" is a bad title for a specification change (CSR) and a release note.
You need to update the title of the PR accordingly. Only you as creator can do that. Thanks.

vish-chan commented:
@tschatzl I've updated the title of the PR and also made code changes as per the review comments. Thanks.

tschatzl reviewed (commented):
~I think this is good~. The change does not compile on Linux due to a missing include to define `INCLUDE_PARALLELGC`

I filed CRs for the renaming of the static members in [JDK-8277372](https://bugs.openjdk.java.net/browse/JDK-8277372)and looking into making card and BOT maximum size dependent on `ObjectAlignmentInBytes` in [JDK-8277373](https://bugs.openjdk.java.net/browse/JDK-8277373)

vish-chan commented:
@tschatzl Yes, the order of include is wrong, sorry about that. I'll add another patch with the correct order.

tschatzl commented:
I will push it through internal testing. Will take a bit.

vish-chan commented:
> I will push it through internal testing. Will take a bit.

@tschatzl Thank you. Regarding  [JDK-8277372](https://bugs.openjdk.java.net/browse/JDK-8277372), do we wait for this PR to integrate and then work on this?

tschatzl commented:
/csr needed

tschatzl commented:
> > I will push it through internal testing. Will take a bit.
> 
> @tschatzl Thank you. Regarding [JDK-8277372](https://bugs.openjdk.java.net/browse/JDK-8277372), do we wait for this PR to integrate and then work on this?

You can certainly start working on this, based on the current version, I'm assuming not too much will change. Integrating this is conditional on CSR approval (which is certain, but takes some time until reviewed).

I do not recommend making a PR already for the JDK-8277372 change yet - these dependent PRs never worked for me :)

albertnetymk commented on src/hotspot/share/gc/shared/cardTable.cpp:
```diff
@@ -428,7 +454,7 @@ MemRegion CardTable::dirty_card_range_after_reset(MemRegion mr,
 
 uintx CardTable::ct_max_alignment_constraint() {
-  return card_size * os::vm_page_size();
+  return GCCardSizeInBytes * os::vm_page_size();
```
The reason why `card_size` can't be used here is that `card_size` haven't been set when this method is called, correct? If so, please put a comment here.

albertnetymk commented on src/hotspot/share/gc/g1/heapRegion.cpp:
```diff
@@ -90,6 +90,9 @@ void HeapRegion::setup_heap_region_size(size_t max_heap_size) {
   GrainWords = GrainBytes >> LogHeapWordSize;
 
+  // Initialize card size
+  CardTable::initialize_card_size();
```
I see that `CardTable::initialize_card_size()` is called at the beginning of `GCArguments::initialize_alignments()` for Serial and Parallel. I wonder if it makes sense to do the same for G1.

vish-chan replied:
The reason for having `CardTable::initialize_card_size()` here in the initial patch was the dependency between card size and G1 heap region size. Since the dependency is no longer there, we can safely move this initialization to `G1Arguments::initialize_alignments`

albertnetymk reviewed (approved):
Just some minor comments/suggestions.

tschatzl commented:
Testing tier1-5 in our CI looks good.

tschatzl commented:
CSR (https://bugs.openjdk.java.net/browse/JDK-8275142) and release note (https://bugs.openjdk.java.net/browse/JDK-8277448) are done. After addressing @albertnetymk 's comments this is ready to go.

vish-chan commented:
> CSR (https://bugs.openjdk.java.net/browse/JDK-8275142) and release note (https://bugs.openjdk.java.net/browse/JDK-8277448) are done. After addressing @albertnetymk 's comments this is ready to go.

@tschatzl I've added a new patch to address @albertnetymk 's comments

tschatzl reviewed (approved):
lgtm.

vish-chan commented:
@tschatzl is there anything else left from my end?

tschatzl commented:
As the bot suggested:

>To flag this PR as ready for integration with the above commit message, type /integrate in a new comment. (Afterwards, your sponsor types /sponsor in a new comment to perform the integration).

Then I'll do the sponsoring (/sponsor) if necesasry.

vish-chan commented:
/integrate

tschatzl commented:
/sponsor
